commands/install.bee.inc

```diff
@@ -86,12 +86,19 @@ function install_bee_callback($arguments, $options) {
 
   // Get database connection string.
   if (!empty($options['db'])) {
-    $db = $options['db'];
+    // Split the passed URL into parts so can encode user, password and db name
+    $db_url = $options['db'];
+    $db_name = rawurlencode(bee_check_plain(bee_after_last('/', $db_url)));
+    $db_username = rawurlencode(bee_check_plain(bee_before(':', bee_after('mysql://', $db_url))));
+    $db_password = rawurlencode(bee_check_plain(bee_between($db_username . ':', '@' . bee_after_last('@', $db_url), $db_url)));
+    $db_host = bee_check_plain(bee_between_last('@', '/', $db_url));
+    // Re-compile the database URL using the sanitized parts
+    $db = "mysql://$db_username:$db_password@$db_host/$db_name";
```

Comment:
Please update this whole file's changes to instead do something similar to what `install.sh` does: https://github.com/backdrop/backdrop/blob/8762dc9369529b6238dda03318394d0f46dc6640/core/scripts/install.sh#L124

Using `parse_url()` should avoid the need for all the helper functions, and `array_map('urldecode', ...)` should simplify things too.

Reply:
@BWPanda - I did try using this, but `parse_url()` is designed for standard hyperlinks and whenever I passed the connection string whole (`--auto` rather than interactive) it failed.  It will not work on a database connection URL with complex passwords.  Take for example `xYK1jxu3$RSOEmL#evqI`; it cannot parse it.  Both the `$` and the `#` cause problems and you can't quote them as that's not valid in a URL either.  I ultimately think `install.sh` is flawed in that it will only accept simple MySQL passwords, whereas MySQL will accept complex passwords and also the GUI installer will handle complex passwords because it takes each element individually.

Reply:
@BWPanda - please can we discuss?

Reply:
Yeah, looks like core still suffers from the same problem: https://github.com/backdrop/backdrop-issues/issues/2904

I think the easiest solution then is to ditch the `db` option and just provide options for each component individually...

Reply:
Done - see comment below.